### DISCUSSION TITLE
Suggest an integration

### DISCUSSION THREAD

**Discussion by oscartbeaumont** [spacedriveapp/spacedrive#95] [Feature Request]
This discussion is a place for the community to recommend cloud integrations for Spacedrive. We would rather you post them as comments here instead of in separate issues to keep the issue tracker clean.

**Note: Everything below is subject to change, so if you disagree with any decisions feel free to start a civil discussion on Discord.**

Our current goal for version one is:
 - [ ] Google Drive
 - [ ] iCloud Drive (Will require a Mac, iPhone or iPad to actually work, thanks to Apple :disappointed:).

We are interested in looking into:
 - [ ] ownCloud/Nextcloud
 - [ ] Object storage (S3, Minio, etc)

Not interesting in supporting at the current stage:

- Syncthing - Syncthing works on its own protocol which would be hard to integrate.

**Please include a link to the cloud service when commenting.**

**Comment by zbraniecki** [comment]
Would you also consider raw AWS S3?

**Reply by oscartbeaumont** [reply]
We will look into it! I wouldn't be surprised if we add it because we may end up using object storage for the Spacedrive hosted service, although obviously, nothing is set in stone at this stage.

**Comment by oscartbeaumont** [comment]
SMB

**Comment by lapidary-master** [comment]
Having integration with MEGA would be super awesome since they are super cheap for large storage.

**Reply by xPolar** [reply]
I believe Mega is planned as a lot of our team members use it. Would be super beneficial to all parties.

**Comment by aiden-dev** [comment]


**Reply by utkubakir** [reply]
Apple Photos, Google Drive, **Dropbox**, **OneDrive** & Mega are planned.

**Comment by ethndotsh** [comment]
AirPort/Time Machine

**Comment by jhodges10** [comment]
Frame.io and Adobe Creative Cloud storage could be good to integrate with for creatives, and is potentially something Adobe would be willing to write or collaborate on.

**Comment by HeavySnowJakarta** [comment]
Some well-known services besides yours provided: (most of them can be easily googled)
- Dropbox
- Onedrive
- Box
- SharePoint 
- IBM Drive
- Yandex Drive
- Baidu Wangpan (Baidu Cloud before) (not easy, may leave for volunteers) [pan.baidu.com] *no English service*
- NutStore (non-webdav would be better)
- Mega
- Pik Pak

Universal remote storation:
- FTP/SFTP
- Windows SMB
- NFS
- WebDAV

Personal storation (usually called nas):
- Synology
- QNAP

And some interesting ideas:
- IPFS
- “WebDrive” based on social networks (eg. Telegram)

**Reply by awidjaja** [reply]
Telegram will be nice (unlimited storage without re-encoding of file/media).

**Comment by ColinLeftwich** [comment]
I don't know how feasible it is but a proton drive integration would be really nice

**Reply by brxken128** [reply]
AFAIK this wouldn't be possible, as Proton don't provide a public API that we could make use of.

It's a shame for sure. I personally use Proton services (and would greatly benefit from a ProtonDrive integration), but they don't like the idea of a public API (in the name of security). It's understandable - an API would be quite complex to develop due to everything they handle being entirely encrypted.

**Comment by PsiACE** [comment]
Wow, I just discovered spacedrive today and it looks just fantastic. I'm excited to see how it can integrate more services.

I haven't gone through the implementation details and I'm not quite sure how to implement an integration. Just thought maybe https://github.com/apache/incubator-opendal might be useful for achieving this goal.

**Comment by ghost** [comment]
great idea @oscartbeaumont

**Comment by Milo123459** [comment]
Falls under S3, but might need more configuration to make it work: https://storj.io

**Comment by sneycampos** [comment]
S3 compatible should be nice, since we can use BackBlaze, Cloudflare R2, etc that is cheap options (Cloudflare R2 charges for egress bandwidth)

**Reply by Zesseth** [reply]
Having BackBlaze B2 through that would be most awesome. 👍

**Comment by kassah** [comment]
I notice all of the above are file storage. I'm still interested in photo storage.

- Google Photos
- ICloud Photos (and disk based Photo libraries)
- Amazon Photos

I'd love to be able to track photo printing services too, but this might not have enough ROI for the effort to implement:
- Walgreens Photo
- Walmart Photo
- Shutterfly

Social services also have their photo type storage:
- Facebook Albums - `https://www.facebook.com/<username>/photos_albums`

**Comment by jihuayu** [comment]
I think the [OpenDAL](https://github.com/apache/incubator-opendal) project might be helpful for this project.
> OpenDAL is a data access layer that allows users to easily and efficiently retrieve data from various storage services in a unified way.

**Comment by ioqbit** [comment]
https://www.resilio.com/individuals/

You can share files on the Internet through a string of string links, without adding friends or user authentication, and conveniently share files between strangers.

**Comment by jcdietrich** [comment]
Support for indexing digital stores that contain downloadable files you own.

- Steam Library
- DriveThruRPG Library
- Itch.io Library
- bundleofholding.com Wizard's Cabinet
  - although most are in DriveThruRPG, some are not
- pnparcade.com
- kobo.com
- amazon.com/amazon.ca/etc

In these cases, I personally have large libraries of files/games that I have access to, but they are all siloed. Indexing across these would be great when trying to find something.

**Reply by iLynxcat** [reply]
This is an interesting idea! We should look into these, though I can't guarantee anything given implementation details/etc. This could be a good use of our upcoming extensions system

**Comment by roaldarbol** [comment]
I'd be really keen to see more encrypted services such as:
- Proton Drive
- Filen
- Ente Photos

I'm sure it's probably not the easiest to integrate... at least Filen is about to have WebDAV support (should land in 2024).

**Reply by roaldarbol** [reply]
What are the technical constraints you are currently working within? Does the service need to have WebDAV or some other interface?

**Reply by roaldarbol** [reply]
Proton has now open sourced their apps (https://github.com/ProtonDriveApps) and I know they're hoping to open up, especially to enable Linux users to interface - so will hopefully be better within the next couple of years. :-)

**Comment by zag2me** [comment]
Cloudflare R2 for me. Sick of using WinSCP 😂

**Comment by oliveiraigorm** [comment]
Any update on SFTP?
